File: Koodi/counter.pl
#!/bin/perl -w

# Count the numbers of Aleph Seq -records in file

use strict;
use utf8;
use List::MoreUtils qw(uniq);

my $beg_time = time;

binmode(STDOUT, ':utf8');

my @recordnumbers;
my $tiedosto = $ARGV[0];
my $linecount = 0;

if( ! defined $tiedosto )
{
  die "Usage: perl counter.pl inputfile\n";
}

open (my $inputfile, '<:utf8', $tiedosto);

while (<$inputfile>)
{
	my $id = substr($_, 0, 9); # Record id
	my $field = substr($_,10, 3); # Field code
	if ($field eq 'FMT')
	{
		push (@recordnumbers, $id);
	}
	$linecount++;	
}

my $totalnumberofrecords = @recordnumbers;
my $uniqueRecords = uniq @recordnumbers;
my $duplicates = ($totalnumberofrecords - $uniqueRecords);

print "The file $tiedosto contains $totalnumberofrecords records ($uniqueRecords unique records, $duplicates duplicates) in $linecount lines.\n";
my $end_time = time;
my $time = ($end_time - $beg_time);
my $minutes = sprintf("%.1f", ($time / 60));
$time = sprintf("%.1f", $time);
print "Processing took $time seconds.\n";
close $inputfile;

File: AlephSequential/counter.py
#!/usr/bin/env python
# -*- coding: utf-8 -*
# Reports the number of distinct Aleph Sequential -records in a file

import sys
import time
import logging

def count():
    start_time = time.time()

    if len(sys.argv) < 2:
    	print("Usage: python counter.py inputfile")
    	sys.exit()
    else:
    	tiedosto = sys.argv[1]

    inputfile = open(tiedosto, "r")

    recordCount = []
    lineCount = 0
    recordCounter = 0
    firstline = inputfile.readline()
    currentid = firstline[:8]
    inputfile.seek(0)

    for line in inputfile:
        id = line[:9]
        field = line[10:13]
        lineCount += 1
        if (id != currentid):
            currentid = id
            recordCount.append(id)
            recordCounter += 1
    numberOfRecords = len(set(recordCount))
    executionTime = (time.time() - start_time)

    print "The file " + tiedosto + " contains " + str(numberOfRecords) + " unique records (" + str(recordCounter) + " records in total) in " + str(lineCount) + " lines."
    print "Processing took %.1f seconds." %  executionTime
    inputfile.close()
    logging.basicConfig(filename='counter.log',level=logging.DEBUG)
    logging.info(' Inputfile: ' + tiedosto)
    logging.info(' Number of unique records in file: ' + str(numberOfRecords))
    logging.info(' Total number of records: ' + str(recordCounter))

if __name__ == '__main__':
    count()